src/lex.rs

```diff
@@ -32,9 +32,18 @@ impl Operator {
     }
     pub fn operate(self, x: f64, y: f64) -> Result<f64, CalcError> {
         if self.token == '/' && y == 0. {
-            Err(CalcError::Math(Math::DivideByZero))
+            return Err(CalcError::Math(Math::DivideByZero));
+        } else if (self.token == '!' && x < 0.0) || (self.token == '!' && x.fract() != 0.0) {
+            return Err(CalcError::Math(Math::OutOfBounds));
+        } else if self.token == '!' && x == 0.0 {
+            // Must return 1 manually as 0..=n where n is 0.0 doesn't work AFAIK.
+            return Ok(1.0);
+        }
```

Comment:
I don't think having `!` at `operate` is the correct place, it should only be checked for `factorial!`.